src/etcd/etcd_master.cc

```diff
@@ -0,0 +1,534 @@
+
+#include "heartbeat_monitor.h"
+#include "src/etcd/etcd_utils.h"
+#include "src/utils.h"
+
+using json = nlohmann::json;
+
+extern "C" {
+aeEventLoop* getEventLoop();
+int getPort();
+}
+
+struct Member {
+  std::string address;
+  std::string port;
+  redisContext* context;
+};
+
+void to_json(json& j, const Member& m) {
+  j = json{{"address", m.address}, {"port", m.port}};
+}
+
+void from_json(const json& j, Member& m) {
+  m.address = j.at("address").get<std::string>();
+  m.port = j.at("port").get<std::string>();
+  m.context = redisConnect(m.address.c_str(), std::stoi(m.port));
+  if (m.context == NULL || m.context->err) {
+    LOG(INFO) << "State in etcd contains unreachable member "
+              << m.address + ":" + m.port;
+  }
+}
+
+EtcdURL etcd_url;
+std::unique_ptr<std::thread> heartbeat_monitor;
+std::shared_ptr<grpc::Channel> etcd_channel;
+std::deque<Member> members;
+
+using Status = leveldb::Status;  // So that it can be easily replaced.
+
+// TODO(zongheng): all callers of SetRole() in this module need to check the
+// return status.
+
+namespace {
+
+// Handle failure for redis module command as caller, hiredis context as callee.
+int ReplyIfFailure(RedisModuleCtx* rm_ctx,
+                   redisContext* ctx,
+                   redisReply* reply) {
+  if (reply == NULL) {
+    return RedisModule_ReplyWithError(rm_ctx, ctx->errstr);
+  } else if (reply->type == REDIS_REPLY_ERROR) {
+    return RedisModule_ReplyWithError(rm_ctx, reply->str);
+  }
+}
+
+}  // anonymous namespace
+
+Status SetRole(redisContext* context,
+               const std::string& role,
+               const std::string& prev_address,
+               const std::string& prev_port,
+               const std::string& next_address,
+               const std::string& next_port,
+               long long* sn_result,
+               long long sn = -1,
+               long long drop_writes = 0) {
+  const std::string sn_string = std::to_string(sn);
+  const std::string drop_writes_string = std::to_string(drop_writes);
+  if (!context->err) {
+    redisReply* reply = reinterpret_cast<redisReply*>(redisCommand(
+        context, "MEMBER.SET_ROLE %s %s %s %s %s %s %s", role.c_str(),
+        prev_address.c_str(), prev_port.c_str(), next_address.c_str(),
+        next_port.c_str(), sn_string.c_str(), drop_writes_string.c_str()));
+    if (reply == NULL) {
+      LOG(INFO) << "reply is NULL, IO error";
+      LOG(INFO) << "error string: " << std::string(context->errstr);
+      return Status::IOError(context->errstr);
+    }
+
+    LOG(INFO) << "Last sequence number is " << reply->integer;
+    *sn_result = reply->integer;
+    freeReplyObject(reply);
+    return Status::OK();
+  } else {
+    LOG(INFO) << "ERROR: SetRole() cannot contact remote node, returning -1";
+    LOG(INFO) << "error string: " << std::string(context->errstr);
+    return Status::IOError(context->errstr);
+  }
+}
+
+grpc::Status WriteChain(const std::deque<Member>& chain) {
```

Comment:
This function is ignoring `chain` and putting in the file-local `members`

Reply:
Fixed